### PR TITLE
implement gitops-run metadata and the remove run command

### LINKED ISSUE #2858: GitOps Run clean
As a user, I get things can go wrong because distributed computing is hard.  Nevertheless, we always need to make sure things are tidy and that includes the usage of `gitops run`.  I would like to have a command that will let me clean up any hanging instances of `gitops run`.  This command should be able to target specific sessions on a cluster or to clean up all sessions.

### Acceptance Criteria

1. When I run the command `gitops remove run` (or some other relevant verb that is consistent with our other commands) then I should see an error stating I need to target a specific session.
2. When I run the command `gitops remove run -n <namespace> <session-id>` then the code should go and find that session on the cluster and clean up the kustomization and bucket and end the process.  
3. When I run the command `gitops remove run -n <namespace> --all-sessions` then the code should go and wreck havoc on the cluster viciously eliminating all instances of gitops run.  This should effectively put the cluster back in "gitops mode". 
4. When I run the command `gitops remove run` in any form and there are no sessions on the cluster at that time I should inform the user that there was nothing to clean up.  This will make them happy knowing that things were already tidy and it was silly of them to run the command.
5. When I run the command `gitops remove run --help` I get some nice helpful information about cleaning these up.

### LINKED ISSUE #2907: GitOps Run: session metadata
For #2898, we need to be able to find our definition of vcluster, and extract a bunch of information about the session.

Already, a vcluster `StatefulSet` is tagged with the label `app: vcluster`. We should also set the following extra label:
 * `app.kubernetes.io/part-of: gitops-run`

We also needs to set these annotations:
 * `run.weave.works/cli-version` to the CLI version
 * `run.weave.works/port-forward` to a comma separated list of ports that are forwarded
 * `run.weave.works/command` to the full command that was run

We can now define a vcluster session as a `StatefulSet` with the two labels `app:vcluster` and `app.kubernetess.io/part-of: gitops-run`, and we can drive the UI with the additional annotations.

### PR DESCRIPTION
**Pull request by chanwit**
Fixes #2858 
Fixes #2907 

How to test this PR
1. Start a GitOps session, for example `gitops beta run ./manifests --port-forward port=8080:5000,resource=svc/app`
2. Wait until it starts the reconciliation loop
3. Open another terminal, and run `killall -9 gitops` to kill all `gitops` process so that the session becomes dangling.
4. `gitops remove run --namespace default --all-sessions` to remove all sessions or you can try specify the session name.

Signed-off-by: Chanwit Kaewkasi <chanwit@weave.works>

### CODE REVIEW

**Comment by opudrovs** [comment]
- First of all, tested the cleanup, it works fine for me, great job! 🎉 

- When running the following command against a clean cluster with `manifests` folder not existing before:

```
gitops beta run ./manifests --port-forward port=8080:5000,resource=svc/app
```

I get `timed out waiting for condition` both in the current and the `main` branch, so it's most likely not related to this PR.

@ozamosi did you any issues like that, can it be related to reported bugs? It cannot be expected behavior, right?

```
olga@macbook-pro podinfo % /Volumes/Macintosh_HD_Data/_job_ww/_src/weave-gitops-2/bin/gitops beta run ./manifests --port-forward port=8080:5000,resource=svc/app
► Checking for a cluster in the kube config ...
► Preparing the cluster for GitOps Run session ...

You can run `gitops beta run --no-session` to disable session management.

If you are running GitOps Run for the first time, it may take a few minutes to download the required images.
GitOps Run session is also required to install Flux components, if it is not installed yet.
You may see Flux installation logs in the next step.

► Checking if Flux is already installed ...
► Getting Flux version ...
⚠️ Flux is not found: error getting Flux labels
Running Flux command: /Users/olga/Library/Caches/.gitops/flux/0.36.0/flux install
✚ generating manifests
✔ manifests build completed
► installing components in flux-system namespace
CustomResourceDefinition/alerts.notification.toolkit.fluxcd.io created
CustomResourceDefinition/buckets.source.toolkit.fluxcd.io created
CustomResourceDefinition/gitrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmcharts.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmreleases.helm.toolkit.fluxcd.io created
CustomResourceDefinition/helmrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/kustomizations.kustomize.toolkit.fluxcd.io created
CustomResourceDefinition/ocirepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/providers.notification.toolkit.fluxcd.io created
CustomResourceDefinition/receivers.notification.toolkit.fluxcd.io created
Namespace/flux-system created
ServiceAccount/flux-system/helm-controller created
ServiceAccount/flux-system/kustomize-controller created
ServiceAccount/flux-system/notification-controller created
ServiceAccount/flux-system/source-controller created
ClusterRole/crd-controller-flux-system created
ClusterRoleBinding/cluster-reconciler-flux-system created
ClusterRoleBinding/crd-controller-flux-system created
Service/flux-system/notification-controller created
Service/flux-system/source-controller created
Service/flux-system/webhook-receiver created
Deployment/flux-system/helm-controller created
Deployment/flux-system/kustomize-controller created
Deployment/flux-system/notification-controller created
Deployment/flux-system/source-controller created
NetworkPolicy/flux-system/allow-egress created
NetworkPolicy/flux-system/allow-scraping created
NetworkPolicy/flux-system/allow-webhooks created
◎ verifying installation
✔ helm-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ notification-controller: deployment ready
✔ source-controller: deployment ready
✔ install finished
► Checking if GitOps Dashboard is already installed ...
Would you like to install the GitOps Dashboard: y
Please enter a password for logging into the dashboard: 
► Creating GitOps Dashboard objects ...
✚ Generating GitOps Dashboard manifests ...
► Creating GitOps Run session run-master-1cf228c6 in namespace default ...

You may see Flux installation logs again, as it is being installed inside the session.

► Waiting for GitOps Run session run-master-1cf228c6 to be ready ...
✔ Session run-master-1cf228c6 is ready
► Connecting to GitOps Run session run-master-1cf228c6 ...
► Checking for a cluster in the kube config ...
► Explicitly allow GitOps Run on run-master-1cf228c6 context
► Checking if Flux is already installed ...
► Getting Flux version ...
⚠️ Flux is not found: error getting Flux labels
Running Flux command: /Users/olga/Library/Caches/.gitops/flux/0.36.0/flux install
✚ generating manifests
✔ manifests build completed
► installing components in flux-system namespace
CustomResourceDefinition/alerts.notification.toolkit.fluxcd.io created
CustomResourceDefinition/buckets.source.toolkit.fluxcd.io created
CustomResourceDefinition/gitrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmcharts.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmreleases.helm.toolkit.fluxcd.io created
CustomResourceDefinition/helmrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/kustomizations.kustomize.toolkit.fluxcd.io created
CustomResourceDefinition/ocirepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/providers.notification.toolkit.fluxcd.io created
CustomResourceDefinition/receivers.notification.toolkit.fluxcd.io created
Namespace/flux-system created
ServiceAccount/flux-system/helm-controller created
ServiceAccount/flux-system/kustomize-controller created
ServiceAccount/flux-system/notification-controller created
ServiceAccount/flux-system/source-controller created
ClusterRole/crd-controller-flux-system created
ClusterRoleBinding/cluster-reconciler-flux-system created
ClusterRoleBinding/crd-controller-flux-system created
Service/flux-system/notification-controller created
Service/flux-system/source-controller created
Service/flux-system/webhook-receiver created
Deployment/flux-system/helm-controller created
Deployment/flux-system/kustomize-controller created
Deployment/flux-system/notification-controller created
Deployment/flux-system/source-controller created
NetworkPolicy/flux-system/allow-egress created
NetworkPolicy/flux-system/allow-scraping created
NetworkPolicy/flux-system/allow-webhooks created
◎ verifying installation
✔ helm-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ notification-controller: deployment ready
✔ source-controller: deployment ready
✔ install finished
► Checking if GitOps Dashboard is already installed ...
► Creating GitOps Dashboard objects ...
✚ Generating GitOps Dashboard manifests ...
► Installing the GitOps Dashboard ...
HelmRelease/flux-system/ww-gitops created
HelmRepository/flux-system/ww-gitops created
✔ GitOps Dashboard has been installed
► Request reconciliation of dashboard (timeout 5m0s) ...
✔ Dashboard reconciliation is done.
► Checking namespace dev-bucket ...
✔ Created namespace dev-bucket
► Checking service dev-bucket/dev-bucket ...
✔ Created service dev-bucket/dev-bucket
► Checking deployment dev-bucket/dev-bucket ...
✔ Created deployment dev-bucket/dev-bucket
► Waiting for deployment dev-bucket to be ready ...
► Port forwarding to pod dev-bucket/dev-bucket-76895f76d-ls4b8 ...
✔ Port forwarding for dev-bucket is ready.
► Port forwarding to pod flux-system/ww-gitops-weave-gitops-577764b7bd-7bm95 ...
✔ Port forwarding for dashboard is ready.
► Checking secret dev-bucket-credentials ...
✔ Created secret dev-bucket-credentials
► Checking bucket source dev-bucket ...
✔ Created source dev-bucket
► Checking Kustomization dev-ks ...
✔ Created ks dev-ks
✔ Setup Bucket Source and Kustomization successfully
◎ Press Ctrl+C to stop GitOps Run ...
► 1 change events detected
► Refreshing bucket dev-bucket ...
...............
► Uploaded 155 files
► Request reconciliation of dev-bucket, and dev-ks (timeout 5m0s) ... 
✗ Error requesting reconciliation: client rate limiter Wait returned an error: context canceled
✗ Error getting pod from specMap: error getting service: client rate limiter Wait returned an error: context canceled, namespaced Name: default/app
► 2 change events detected
► Refreshing bucket dev-bucket ...
...............
► Uploaded 155 files
► Request reconciliation of dev-bucket, and dev-ks (timeout 5m0s) ... 
✗ Error requesting reconciliation: timed out waiting for the condition
✗ Error getting pod from specMap: error getting service: services "app" not found, namespaced Name: default/app
^C

► Deleting GitOps Run session run-master-1cf228c6 ...
^C
olga@macbook-pro podinfo % 
olga@macbook-pro podinfo % 
```

I tested it several times, also with my standard podinfo portforwarding option and without any portforwarding, the same. The `manifests` folder with the following kustomization.yaml file in it is created successfully:

```
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources: [] # 👋 Start adding the resources you want to sync here
```

File: cmd/gitops/remove/run/cmd.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	cmdFlags.BoolVar(&flags.AllSessions, "all-sessions", false, "Remove all GitOps Run sessions")
+
+	kubeConfigArgs = run.GetKubeConfigArgs()
+
+	kubeConfigArgs.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func getKubeClient(cmd *cobra.Command, args []string) (*kube.KubeHTTP, *rest.Config, error) {
```
Comment (opudrovs) [inline cmd/gitops/remove/run/cmd.go:66]: 💡  Is there any difference between this function and the `getKubeClient` function here?
https://github.com/weaveworks/weave-gitops/blob/session-remove/cmd/gitops/beta/run/cmd.go#L181

I only see the following log message removed `log.Actionf("Checking for a cluster in the kube config ...")` in the remove cmd.go version comparing to the run cmd.go version. Can we just remove this log message and combine these two functions?

**Review by opudrovs — APPROVED** [review]

File: cmd/gitops/remove/run/cmd.go (hunk 1 above)
Comment (chanwit) [reply cmd/gitops/remove/run/cmd.go:66]: Yes, I need to refactor it somehow. Haven't got a good idea how should I do it yet.

**Comment by chanwit** [comment]
I tried to refactor it but I think it's better for the refactoring to be in a separate ticket.

**Comment by chanwit** [comment]
We will follow up the refactoring in #2946

**Pull request merged**